### ISSUE TITLE
Bug: Failed to open stream: No such file or directory

### ISSUE THREAD

**Issue by afairhurst** [heavymetalavo/craft-aialttext#1]
Hi,

I'm trialing your Ai Alt Text plugin. 

- I'm currently testing in a local dev environment using OrbStack.
- Images are hosted at DigitalOcean Spaces
- Craft CMS Pro 5.6.13

When I try to generate Ai Alt text, it fails with this error:
`file_get_contents(IMG_1760-2-bw.jpg): Failed to open stream: No such file or directory`

Please advise.

Thank you,

Andrew

```
2025-03-28 12:01:55 [queue.INFO] [craft\queue\QueueLogBehavior::beforeExec]  [32272] Generating alt text for IMG_1760-2-bw.jpg, Element ID: 4169 (attempt: 1, pid: 19) - Started {"memory":20192648} 
2025-03-28 12:01:55 [queue.ERROR] [craft\queue\QueueLogBehavior::afterError]  [32272] Generating alt text for IMG_1760-2-bw.jpg, Element ID: 4169 (attempt: 1, pid: 19) - Error (time: 0.033s): file_get_contents(IMG_1760-2-bw.jpg): Failed to open stream: No such file or directory {"memory":23068248} 
2025-03-28 12:01:55 [queue.ERROR] [yii\base\ErrorException:2] yii\base\ErrorException: file_get_contents(IMG_1760-2-bw.jpg): Failed to open stream: No such file or directory in /var/www/project/vendor/heavymetalavo/craft-aialttext/src/services/AiAltTextService.php:74

```

**Comment by afairhurst** [comment]
I've just tried testing the plugin with locally hosted assets. It's working about 50% of the time; sometimes failing, sometimes, not. When failures occur, I'm seeing this error:

`Empty alt text generated for asset: IMG_2367.jpg`

From the Queue Log:

```
2025-03-28 14:15:30 [queue.INFO] [craft\queue\QueueLogBehavior::beforeExec]  [32741] Generating alt text for IMG_2367.jpg, Element ID: 2065 (attempt: 1, pid: 19) - Started {"memory":20194688} 
2025-03-28 14:15:30 [queue.WARNING] [heavymetalavo\craftaialttext\services\OpenAiService::isUrlAccessible] URL accessibility check failed: cURL error 3: URL rejected: No host part in the URL (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for /assets/uploads/Alt-Text/IMG_2367.jpg {"memory":24934600} 
2025-03-28 14:15:30 [queue.WARNING] [heavymetalavo\craftaialttext\services\OpenAiService::generateAltText] Asset URL is not accessible remotely: /assets/uploads/Alt-Text/IMG_2367.jpg {"memory":24894456} 
2025-03-28 14:15:30 [queue.ERROR] [heavymetalavo\craftaialttext\services\OpenAiService::generateAltText] Failed to generate alt text: file_get_contents(assets/uploads/Alt-Text/IMG_2367.jpg): Failed to open stream: No such file or directory {"memory":24922544} 
2025-03-28 14:15:30 [queue.ERROR] [craft\queue\QueueLogBehavior::afterError]  [32741] Generating alt text for IMG_2367.jpg, Element ID: 2065 (attempt: 1, pid: 19) - Error (time: 0.104s): Empty alt text generated for asset: IMG_2367.jpg {"memory":24920992} 
2025-03-28 14:15:30 [queue.ERROR] [Exception] Exception: Empty alt text generated for asset: IMG_2367.jpg in /var/www/project/vendor/heavymetalavo/craft-aialttext/src/services/AiAltTextService.php:85
Stack trace:
#0 /var/www/project/vendor/heavymetalavo/craft-aialttext/src/jobs/GenerateAiAltText.php(37): heavymetalavo\craftaialttext\services\AiAltTextService->generateAltText(Object(craft\elements\Asset))
#1 /var/www/project/vendor/yiisoft/yii2-queue/src/Queue.php(243): heavymetalavo\craftaialttext\jobs\GenerateAiAltText->execute(Object(craft\queue\Queue))
#2 /var/www/project/vendor/yiisoft/yii2-queue/src/cli/Queue.php(162): yii\queue\Queue->handleMessage('32741', 'O:51:"heavymeta...', '300', '1')
#3 /var/www/project/vendor/yiisoft/yii2-queue/src/cli/Command.php(146): yii\queue\cli\Queue->execute('32741', 'O:51:"heavymeta...', '300', '1', '19')
#4 [internal function]: yii\queue\cli\Command->actionExec('32741', '300', '1', '19')
#5 /var/www/project/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#6 /var/www/project/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#7 /var/www/project/vendor/yiisoft/yii2/console/Controller.php(180): yii\base\Controller->runAction('exec', Array)
#8 /var/www/project/vendor/craftcms/cms/src/console/ControllerTrait.php(88): yii\console\Controller->runAction('exec', Array)
#9 /var/www/project/vendor/yiisoft/yii2/base/Module.php(552): craft\queue\Command->runAction('exec', Array)
#10 /var/www/project/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('queue/exec', Array)
#11 /var/www/project/vendor/craftcms/cms/src/console/Application.php(91): yii\console\Application->runAction('queue/exec', Array)
#12 /var/www/project/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('queue/exec', Array)
#13 /var/www/project/vendor/craftcms/cms/src/console/Application.php(122): yii\console\Application->handleRequest(Object(craft\console\Request))
#14 /var/www/project/vendor/yiisoft/yii2/base/Application.php(384): craft\console\Application->handleRequest(Object(craft\console\Request))
#15 /var/www/project/craft(13): yii\base\Application->run()
#16 {main} {"memory":24921208,"exception":"[object] (Exception(code: 0): Empty alt text generated for asset: IMG_2367.jpg at /var/www/project/vendor/heavymetalavo/craft-aialttext/src/services/AiAltTextService.php:85)"} 
2025-03-28 14:15:30 [queue.INFO] [yii\db\Connection::open] Opening DB connection: mysql:host=mysql;dbname=project;port=3306 {"memory":24928272} 
2025-03-28 14:15:30 [queue.ERROR] [Exception] Exception: Empty alt text generated for asset: IMG_2367.jpg in /var/www/project/vendor/heavymetalavo/craft-aialttext/src/services/AiAltTextService.php:85
Stack trace:
#0 /var/www/project/vendor/heavymetalavo/craft-aialttext/src/jobs/GenerateAiAltText.php(37): heavymetalavo\craftaialttext\services\AiAltTextService->generateAltText(Object(craft\elements\Asset))
#1 /var/www/project/vendor/yiisoft/yii2-queue/src/Queue.php(243): heavymetalavo\craftaialttext\jobs\GenerateAiAltText->execute(Object(craft\queue\Queue))
#2 /var/www/project/vendor/yiisoft/yii2-queue/src/cli/Queue.php(162): yii\queue\Queue->handleMessage('32741', 'O:51:"heavymeta...', '300', '1')
#3 /var/www/project/vendor/yiisoft/yii2-queue/src/cli/Command.php(146): yii\queue\cli\Queue->execute('32741', 'O:51:"heavymeta...', '300', '1', '19')
#4 [internal function]: yii\queue\cli\Command->actionExec('32741', '300', '1', '19')
#5 /var/www/project/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#6 /var/www/project/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#7 /var/www/project/vendor/yiisoft/yii2/console/Controller.php(180): yii\base\Controller->runAction('exec', Array)
#8 /var/www/project/vendor/craftcms/cms/src/console/ControllerTrait.php(88): yii\console\Controller->runAction('exec', Array)
#9 /var/www/project/vendor/yiisoft/yii2/base/Module.php(552): craft\queue\Command->runAction('exec', Array)
#10 /var/www/project/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('queue/exec', Array)
#11 /var/www/project/vendor/craftcms/cms/src/console/Application.php(91): yii\console\Application->runAction('queue/exec', Array)
#12 /var/www/project/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('queue/exec', Array)
#13 /var/www/project/vendor/craftcms/cms/src/console/Application.php(122): yii\console\Application->handleRequest(Object(craft\console\Request))
#14 /var/www/project/vendor/yiisoft/yii2/base/Application.php(384): craft\console\Application->handleRequest(Object(craft\console\Request))
#15 /var/www/project/craft(13): yii\base\Application->run()
#16 {main} {"memory":24970032,"exception":"[object] (Exception(code: 0): Empty alt text generated for asset: IMG_2367.jpg at /var/www/project/vendor/heavymetalavo/craft-aialttext/src/services/AiAltTextService.php:85)"} 
```


Thank you

**Comment by joepagan** [comment]
Hi @afairhurst thanks for reporting the issue, can you confirm which version you tested? 
I was working on improving this about the time you reported this and released 1.0.7 which may have resolved pathing for local assets.

**Comment by afairhurst** [comment]
Hi Joe,

At first I was using  1.0.6, but I upgraded to  1.0.7 when I saw it was available. I've since upgraded to v1.2.0.

Error remains:
Failed to open stream: No such file or directory

Locally hosted assets work fine. The issue is only with assets hosted with DigitalOcean spaces.

Thank you,

Andrew

**Comment by joepagan** [comment]
Hey Andrew sorry completely missed that you were on DO I will try replicate that.

Are the images which fail available on a public URL or privatised?

**Comment by afairhurst** [comment]
Thank you. All are publically accessible.

**Comment by joepagan** [comment]
Hey @afairhurst I presume you are using [vaersaagod/dospaces](https://github.com/vaersaagod/dospaces) to connect to your do spaces bucket? 

I managed to replicate this issue with very similar logs to what you shared, but only where this setting is off:
<img width="308" alt="Image" src="https://github.com/user-attachments/assets/b47c355e-7ba7-49e8-af5f-e60f85c5a0ea" />

Maybe it's possible the asset's are technically publicly available but the setting was off for you?

That said, with the setting off I was able to make a fix to support this and I've released [v1.2.1](https://github.com/heavymetalavo/craft-aialttext/releases/tag/v1.2.1) 

Can you give this new version a try and see if that works for you?

**Comment by afairhurst** [comment]
Hi Joe,

Thank you so much for taking the time to look into this.

Maybe my setup is unusual, however I have two filesystems, which use vaersaagod/dospaces plugin to connect to the DO Spaces bucket. The first is for the original images, which aren't publically accessible, whilst the second contains transformed assets.

**Originals**
<img width="1115" alt="Image" src="https://github.com/user-attachments/assets/c43f4b9b-ac38-471e-b7d3-8898a34a4c5c" />

**Transformed**
<img width="1117" alt="Image" src="https://github.com/user-attachments/assets/a2fb3514-5647-452b-b56c-79f5f3836bb6" />

**Example Volume**
<img width="1122" alt="Image" src="https://github.com/user-attachments/assets/178d5f1e-b084-4548-ad4b-62313d79f8c0" />

I'll try the new version now.

Thank you,

Andrew

**Comment by afairhurst** [comment]
Brilliant; The latest update seems to be working perfectly now. I'll let you know if anything arises with further testing, but so far, so good. Thank you!

**Comment by joepagan** [comment]
Woop! @afairhurst thanks for reporting back, glad to hear you are sorted! 

As you say, I think the issue was that your originals bucket is technically where the assets are stored which is privatised which is what the plugin is using to send off to OpenAI. 

I'll go ahead and close this but of course feel free to comment if it comes up again and I'll take another look.

**Comment by joepagan** [comment]
@afairhurst if you could leave a review on the plugin store I'd really appreciate it 🙏

**Comment by joepagan** [comment]
Thanks @afairhurst much love!
